ports/stm32/machine_i2c.c

```diff
@@ -37,38 +37,45 @@
 
 STATIC const mp_obj_type_t machine_hard_i2c_type;
 
+#define I2C_POLL_TIMEOUT_US_DEFAULT (50000) // 50ms
+
 #if defined(STM32F0) || defined(STM32F4) || defined(STM32F7)
 
 typedef struct _machine_hard_i2c_obj_t {
     mp_obj_base_t base;
     i2c_t *i2c;
     mp_hal_pin_obj_t scl;
     mp_hal_pin_obj_t sda;
+    uint32_t *us_timeout;
 } machine_hard_i2c_obj_t;
 
+STATIC uint32_t machine_hard_i2c_timeout[];
+
 STATIC const machine_hard_i2c_obj_t machine_hard_i2c_obj[] = {
     #if defined(MICROPY_HW_I2C1_SCL)
-    {{&machine_hard_i2c_type}, I2C1, MICROPY_HW_I2C1_SCL, MICROPY_HW_I2C1_SDA},
+    {{&machine_hard_i2c_type}, I2C1, MICROPY_HW_I2C1_SCL, MICROPY_HW_I2C1_SDA, &machine_hard_i2c_timeout[0]},
     #else
-    {{NULL}, NULL, NULL, NULL},
+    {{NULL}, NULL, NULL, NULL, NULL},
     #endif
     #if defined(MICROPY_HW_I2C2_SCL)
-    {{&machine_hard_i2c_type}, I2C2, MICROPY_HW_I2C2_SCL, MICROPY_HW_I2C2_SDA},
+    {{&machine_hard_i2c_type}, I2C2, MICROPY_HW_I2C2_SCL, MICROPY_HW_I2C2_SDA, &machine_hard_i2c_timeout[1]},
     #else
-    {{NULL}, NULL, NULL, NULL},
+    {{NULL}, NULL, NULL, NULL, NULL},
     #endif
     #if defined(MICROPY_HW_I2C3_SCL)
-    {{&machine_hard_i2c_type}, I2C3, MICROPY_HW_I2C3_SCL, MICROPY_HW_I2C3_SDA},
+    {{&machine_hard_i2c_type}, I2C3, MICROPY_HW_I2C3_SCL, MICROPY_HW_I2C3_SDA, &machine_hard_i2c_timeout[2]},
     #else
-    {{NULL}, NULL, NULL, NULL},
+    {{NULL}, NULL, NULL, NULL, NULL},
     #endif
     #if defined(MICROPY_HW_I2C4_SCL)
-    {{&machine_hard_i2c_type}, I2C4, MICROPY_HW_I2C4_SCL, MICROPY_HW_I2C4_SDA},
+    {{&machine_hard_i2c_type}, I2C4, MICROPY_HW_I2C4_SCL, MICROPY_HW_I2C4_SDA, &machine_hard_i2c_timeout[3]},
     #else
-    {{NULL}, NULL, NULL, NULL},
+    {{NULL}, NULL, NULL, NULL, NULL},
     #endif
 };
 
+STATIC uint32_t machine_hard_i2c_timeout[sizeof(machine_hard_i2c_obj)/sizeof(machine_hard_i2c_obj_t)] = {I2C_POLL_TIMEOUT_US_DEFAULT};
```

Comment:
I'm not sure all compilers will like this use of `sizeof()` here... is it legal?  Note there's the macro `MP_ARRAY_SIZE()` for this.

Also, I don't think it's necessary to initialise the value, it can just stay "blank", ie in the BSS.  The timeout should always be populated upon creation of an I2C object.  (And being in the BSS is slightly more optimal, no code space taken, initing it at boot is faster.)